articles/contract-migration-guide.md

```diff
@@ -0,0 +1,239 @@
+
+Casper 2.0 introduces changes that require adaptation of your existing Casper 1.x smart contracts. This guide outlines the crucial updates and provides a step-by-step approach for a seamless upgrade, using the CEP-18 contract as a reference.
+
+## Key Changes in Casper 2.0 and Their Implications
+
+1. **Key Representations:**
+   - **`Key::AddressableEntity`:**
+     - In Casper 2.0, `Key::AddressableEntity` is the unifed representation for system, account and contract keys. The `EntityKindTag`  specifies their type:
+
+        ```rust
+        pub enum EntityKindTag {
+            /// `EntityKind::System` variant.
+            System = 0,
+            /// `EntityKind::Account` variant.
+            Account = 1,
+            /// `EntityKind::SmartContract` variant.
+            SmartContract = 2,
+        }
+        ``` 
+     - Ensure you update all key references in your contract to use this new format.
+
+2. **Entry Point Declarations:**
+    - **`EntryPointType`:**
+        - **EntryPointType::Caller:** This is the main entry point type used in Casper 2.0. It signifies that the entry point will be executed using the calling entity's context (the account that initiated the transaction)
+        - **EntryPointType::Called:** This type is used for entry points that are specifically designed to be called by other contracts. These entry points run in the context of the contract being called. This is functionally similar to the previous `Contract` type in Casper 1.x.
+        - **EntryPointType::Factory:** This is a new entry point type in Casper 2.0. It's specifically designed for contracts that create new contracts (like factory contracts). These entry points also run in the context of the called contract.
+        - The `StoredSession` entry point type is no longer available in Casper 2.0 (`EntryPointType::Caller` does not provide that functionality).
+
+    - **`EntryPointPayment`:**
+        - `EntryPointPayment::Caller`: The caller of the entry point pays for the execution cost.
+        - `EntryPointPayment::SelfOnly`: The contract itself pays for its own execution but not for any subsequent contract calls.
+        - `EntryPointPayment::SelfOnward`: The contract pays for its own execution and any subsequent contract calls it makes.
+
+    Example (CEP-18 `symbol` Entry Point):
+
+    ```rust
+    pub fn symbol() -> EntryPoint {
+        EntryPoint::new(
+            String::from(SYMBOL_ENTRY_POINT_NAME),
+            Vec::new(),
+            String::cl_type(),
+            EntryPointAccess::Public,
+            EntryPointType::Called,  // NEW
+            casper_types::EntryPointPayment::Caller, // NEW
+        )
+    }
+    ```
+   
+3. **Event Handling:**
+   - **Native Events and Message Topics:**
+     - Casper 2.0 prioritizes native event emission through message topics, offering flexibility and better integration.
+     - Utilize `runtime::emit_message` to send events on specific topics.
+   - **CES (Casper Event Standard):**
```

Comment:
In case this is a doc for overall condor migration, CES should not be here. That is an implementation detail per contract and is not relevant for general condor readiness.

Reply:
I kept it here as a major change for smart contract developers with a note `While not mandatory, switching to native events is recommended due to their cost-efficiency`. I removed the info about events from the `Upgrade steps` section